src/lib/include/x86_gintrin.h

```diff
@@ -0,0 +1,232 @@
+// Define generic functions that map to SSE/AVX intrinsics.
+//
+// For example the intrinsic for floating-point addition comes in four
+// versions: _mm_add_ps, _mm_add_pd, _mm256_add_ps, _mm256_add_pd. To save us
+// writing the same code many times, this file defines generic intrinsics. The
+// intrinsic for add becomes add_p, and the correct overload is chosen by the
+// compiler.
+//
+// This is a bit ugly because we're tricking the preprocessor into
+// automatically writing four versions of every function for us.
+
+#ifndef DISTOPIA_X86_GINTRIN_H
+#define DISTOPIA_X86_GINTRIN_H
+
+#include "arch_config.h"
+#ifdef DISTOPIA_X86_SSE4_1
+
+#include <immintrin.h>
+
+namespace {
+
+#ifdef DISTOPIA_GCC
+// This well-meaning warning from GCC is irrelevant here.
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wignored-attributes"
+#endif
+
+// Map native vector type to corresponding scalar type.
+template<typename vectorT> struct ScalarTFromVecTStruct;
+template<> struct ScalarTFromVecTStruct<__m128> { using type = float; };
+template<> struct ScalarTFromVecTStruct<__m128d> { using type = double; };
+template<> struct ScalarTFromVecTStruct<__m256> { using type = float; };
+template<> struct ScalarTFromVecTStruct<__m256d> { using type = double; };
+template<typename vectorT>
+using ScalarTFromVecT = typename ScalarTFromVecTStruct<vectorT>::type;
+
+// These are intrinsics: ensure they are always inlined.
+#ifdef __GNUC__
+  #define DistopiaGintrinAttrs inline __attribute__((__always_inline__))
+#else
+  #define DistopiaGintrinAttrs inline
+#endif
+
+// DistopiaGintrinUnaryFromScalar is templated and must have a non-specialized
+// prototype.
+#define DistopiaGintrinUnaryFromScalarPrototype(Fname, Argqual) \
+  template<typename vectorT> \
+  DistopiaGintrinAttrs vectorT Fname ## _p(ScalarTFromVecT<vectorT> Argqual a);
```

Comment:
@nbgl A bit more explanation as to what these defines actually do would be amazing. Its a bit of a mind bend. :)

Reply:
I’ve added some examples. Are they helpful?